### DISCUSSION TITLE
Line item splitting and removal

### DISCUSSION THREAD

**Discussion by BrandoCommando** [Shopify/function-examples#261] [Feedback: Cart Transform API]
My proposition is simple: allow line item splitting/removal and attribute changes via Cart Transform functions. I realize the ["migration guide" from scripts to functions](https://shopify.dev/docs/api/functions/reference/product-discounts#migrating-from-scripts-to-functions) says to use checkout extensions to apply attribute changes, but that would only work for very specific uses.

Here's my use case:
- Many products have paid "upgrades" available. We handle this by adding 2 separate variants to the cart, and using attributes to keep track of their relationships.
- Many products have "mix and match" bundles that span between variants, and sometimes separate products. We also use attributes to track those relationships.

Both of these things are much simpler to manage if all line items have a quantity of 1. Our Line Item Script currently does the following:

1. Ensure each line item has quantity one. If one is greater, split it into more, assigning each line a separate attribute.
2. Make sure upgrade line items have an upgrade attribute that matches another general product upgrade attribute. If not, remove the line item.
3. Iterate the line items, looking for "bundles" that match several criteria. If the bundle quantity is met, lower the price of each item the same amount, and update their bundle attributes to be the same.

When first implementing our upgrade/bundling logic, we did it all on the front-end use the cart Ajax API. However, this was prone to errors as well as client-side issues (timeouts probably?) resulting in floating upgrades and un-matched attributes. So then we switched over to Scripts, which solved those problems. Now that Scripts is going to be deprecated with Checkout.liquid, we need a suitable replacement. We need to be able to remove invalid line items, and split lines in order to ensure quantities of one.

**Comment by nboliver-ventureweb** [comment]
Agree, this should be supported. Also think adding an item from a function should be made possible.
